sycl/include/sycl/detail/spirv.hpp

```diff
@@ -775,217 +803,322 @@ inline uint32_t membermask() {
+  if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                    GroupT> &&
+                detail::is_vec<T>::value) {
+    // Temporary work-around due to a bug in IGC.
+    // TODO: Remove when IGC bug is fixed.
+    T result;
+    for (int s = 0; s < x.size(); ++s)
+      result[s] = ShuffleDown(g, x[s], delta);
+    return result;
+  } else if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                           GroupT>) {
+    // To avoid overflow we adjust the delta down by the size of the group, if
+    // it is larger than the size of the group. Since the local ID cannot be
+    // larger than the size of the group, it cannot overflow when we add them
+    // prior to wrapping the target ID.
+    id<1> TargetLocalId = g.get_local_id();
+    size_t ItemsInGroup = g.get_local_linear_range();
+    uint32_t AdjustedDelta =
+        delta > ItemsInGroup ? delta - ItemsInGroup : delta;
+    TargetLocalId[0] = (TargetLocalId[0] + AdjustedDelta) % ItemsInGroup;
+    uint32_t TargetId = MapShuffleID(g, TargetLocalId);
+    return __spirv_GroupNonUniformShuffle(group_scope<GroupT>::value,
+                                          convertToOpenCLType(x), TargetId);
+  } else {
+    // Subgroup.
+    return __spirv_SubgroupShuffleDownINTEL(convertToOpenCLType(x),
+                                            convertToOpenCLType(x), delta);
+  }
+  if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                    GroupT>) {
+    return __nvvm_shfl_sync_down_i32(detail::ExtractMask(detail::GetMask(g))[0],
+                                     x, delta, 0x1f);
+  } else {
+    return __nvvm_shfl_sync_down_i32(membermask(), x, delta, 0x1f);
+  }
 #endif
 }
 
-template <typename T>
-EnableIfNativeShuffle<T> SubgroupShuffleUp(T x, uint32_t delta) {
+template <typename GroupT, typename T>
+EnableIfNativeShuffle<T> ShuffleUp(GroupT g, T x, uint32_t delta) {
 #ifndef __NVPTX__
-  return __spirv_SubgroupShuffleUpINTEL(convertToOpenCLType(x),
-                                        convertToOpenCLType(x), delta);
+  if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                    GroupT> &&
+                detail::is_vec<T>::value) {
+    // Temporary work-around due to a bug in IGC.
+    // TODO: Remove when IGC bug is fixed.
+    T result;
+    for (int s = 0; s < x.size(); ++s)
+      result[s] = ShuffleUp(g, x[s], delta);
+    return result;
+  } else if constexpr (ext::oneapi::experimental::is_user_constructed_group_v<
+                           GroupT>) {
+    id<1> TargetLocalId = g.get_local_id();
+    size_t ItemsInGroup = g.get_local_linear_range();
+    uint32_t WrappedDelta = delta % ItemsInGroup;
+    // Avoid underflow.
+    if (TargetLocalId[0] < WrappedDelta)
+      TargetLocalId[0] = ItemsInGroup - (WrappedDelta - TargetLocalId[0]);
+    else
+      TargetLocalId[0] -= WrappedDelta;
```

Comment:
Nit: Same comment as for the overflow case.  Could we say `uint32_t AdjustedDelta = delta >= ItemsInGroup ?  0 : delta;`  as well?

Reply:
See previous comment. We can discuss further in there.